README
README

This project implements pair-wise (local and global) alignments for
sequences that consist of alphabets that are organised in a
hierarchical classification scheme. The hierarchical classification of
the alphabet affords us the ability to query a path distance between
any two symobls in the alphabet, which incorporates more detail than
traditional binary match/mismatch queries. This path distance
inherently revolves around knowing the least common ancestor of the
two symbols. In order to facilitate efficient lookups of the LCAs we
computed all pair-wise LCAs offline using the lca application (also on
github) [1].

The inputs to the application have to satisfy the following format:

1) *Euler Levels*: the euler levels are the result of computing the
Euler circuit of the hierarchical classification system [2]. This is
just a row vector with the levels in the classification tree.

2) *Euler Positions*: the positions of the first occurrence of a
symbol in the Euler circuit is also computed by the the application in
[2]. This file gives the row vector of the indices into the euler
circuit.

3) *LCA*: the least common ancestor lookup table is computed using
[1]. The file defines 3 columns separated by commas of the 2 symbols
and the corresponding LCA.

4) *Set 1/2*: A file of comma-separated sequences defined over the
alphabet.

The output is a file with the pair-wise normalised similarity score
where the pair constitutes a comparison between any sequence of set 1
with any sequence of set 2.


CONFIGURATION

The code is implemented in C++ using the boost libraries.

OpenMP can be enabled using the --enable-openmp flag at the
configuration step.


EXECUTION

General Configuration:
  --help                produce help message
  --version             show the version

I/O Configuration:
  --results arg (=./results) results directory.
  --euler_levels arg         Filename of the vertex levels in the Euler 
                             Circuit.
  --euler_positions arg      Filename of the vertex positions in the Euler 
                             Circuit.
  --lca arg                  Filename with the LCAs computed offline.
  --set_1 arg                Filename of the source set.
  --set_2 arg                Filename of the target set.

Algorithm Configuration:
  --alg arg (=1)             Algorithm: 1 - local alignment, 2 - global alignment.
  --scores arg (=0)          Compute just alignment scores, no backtracking.
  --gap_penalty arg (=1.33)  Gap penalty for the alignments.


[1] https://github.com/dahlem/lca
[2] https://github.com/dahlem/Euler-Circuit


ChangeLog
* [0.5] - 2015-05-05
** Changed
 - Implemented LCA using the euler circuit and level arrays


src/main/main.cc
// Copyright (C) 2015 Dominik Dahlem <Dominik.Dahlem@gmail.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

#if HAVE_CONFIG_H
# include <config.h>
#endif

#ifndef __STDC_CONSTANT_MACROS
# define __STDC_CONSTANT_MACROS
#endif /* __STDC_CONSTANT_MACROS */

#ifndef NDEBUG
# include <iterator>
#endif /* NDEBUG */

#include <algorithm>
#include <fstream>
#include <iostream>
#include <vector>

#include <boost/cstdint.hpp>
#include <boost/flyweight.hpp>
#include <boost/lexical_cast.hpp>
#include <boost/tokenizer.hpp>

#include <boost/algorithm/string/split.hpp>
#include <boost/algorithm/string/trim.hpp>

#include "CL.hh"
#include "Types.hh"
#include "TreePathSimilarityMeasure.hh"

#include "AbstractDistanceMeasure.hh"
#include "SimilarityAlgorithm.hh"
#include "NW.hh"
#include "SW.hh"


typedef boost::tokenizer <boost::escaped_list_separator <char> > Tokenizer;

static common::Symbol::initializer fw_symbol_init;


namespace std
{

std::ostream& operator<<(std::ostream &os, const common::StringIntMap::value_type &v) {
  return os << v.first << ": " << v.second;
}

std::ostream& operator<<(std::ostream &os, const common::StrStrMap::value_type &v) {
  return os << std::get<0>(v.first) << ", " << std::get<1>(v.first) << ": " << v.second;
}

std::ostream& operator<<(std::ostream &os, const common::StringVec &v) {
  std::copy(v.begin(), v.end(), std::ostream_iterator<common::Symbol>(os, "\t"));
  return os;
}

}

int main(int argc, char *argv[])
{
  args_t args;
  CL cl;

  if (cl.parse(argc, argv, args)) {
    return EXIT_SUCCESS;
  }

  // parse the euler levels
  std::ifstream eulerLevelsFile;
  eulerLevelsFile.open(args.euler_levels.c_str());
  if (!eulerLevelsFile.is_open()) {
    std::cerr << "Could not open file: " << args.euler_levels << std::endl;
    return EXIT_FAILURE;
  }

#ifndef NDEBUG
  std::cout << "Reading the Euler Levels..." << std::endl;
#endif /* NDEBUG */

  common::IntVec euler_levels;
  std::string line;
  while (!eulerLevelsFile.eof()) {
    std::getline(eulerLevelsFile, line);
#ifndef NDEBUG
    std::cout << "Read line: " << line << std::endl;
#endif /* NDEBUG */

    if (line != "") {
      boost::algorithm::trim(line);
      euler_levels.push_back(boost::lexical_cast<boost::uint32_t>(line));
    }
  }
  eulerLevelsFile.close();


  // parse the euler positions
  std::ifstream eulerPositionsFile;
  eulerPositionsFile.open(args.euler_positions.c_str());
  if (!eulerPositionsFile.is_open()) {
    std::cerr << "Could not open file: " << args.euler_positions << std::endl;
    return EXIT_FAILURE;
  }

#ifndef NDEBUG
  std::cout << "Reading the Euler Positions..." << std::endl;
#endif /* NDEBUG */

  common::StringIntMap euler_positions;
  while (!eulerPositionsFile.eof()) {
    std::getline(eulerPositionsFile, line);
#ifndef NDEBUG
    std::cout << "Read line: " << line << std::endl;
#endif /* NDEBUG */

    if (line != "") {
      common::StringVec tokens;
      boost::split(tokens, line, boost::is_any_of(","), boost::token_compress_on);
      if (tokens.size() != 2) {
        eulerPositionsFile.close();
        std::cerr << "Each line of the euler positions file can only contain two tokens: <key>,<value>!" << std::endl;
        return EXIT_FAILURE;
      } else {
        if (euler_positions.find(tokens[0]) != euler_positions.end()) {
          eulerPositionsFile.close();
          std::cerr << "The Euler Positions cannot contain duplicates: " << tokens[0] << std::endl;
          return EXIT_FAILURE;
        } else {
          euler_positions[tokens[0]] = boost::lexical_cast<boost::uint32_t>(tokens[1]);
        }
      }
    }
  }
  eulerPositionsFile.close();

#ifndef NDEBUG
  std::cout << std::endl << "Euler Levels:  ";
  std::copy(euler_levels.begin(), euler_levels.end(), std::ostream_iterator<boost::uint32_t>(std::cout, "\t"));
  std::cout << std::endl;
  std::cout << std::endl << "Euler Positions:  ";
  std::copy(euler_positions.begin(), euler_positions.end(), std::ostream_iterator<common::StringIntMap::value_type>(std::cout, "\t"));
  std::cout << std::endl;
#endif /* NDEBUG */

  std::ifstream lcaFile;
  lcaFile.open(args.lca.c_str());
  if (!lcaFile.is_open()) {
    std::cerr << "Could not open file: " << args.lca << std::endl;
    return EXIT_FAILURE;
  }

#ifndef NDEBUG
  std::cout << "Reading the lca..." << std::endl;
#endif /* NDEBUG */

  common::StrStrMap lcas;
  while (!lcaFile.eof()) {
    std::getline(lcaFile, line);
#ifndef NDEBUG
    std::cout << "Read line: " << line << std::endl;
#endif /* NDEBUG */

    if (line != "") {
      common::StringVec tokens;
      boost::split(tokens, line, boost::is_any_of(","), boost::token_compress_on);
      if (tokens.size() != 3) {
        std::cerr << "The LCA requires three tokens, got: " << tokens.size() << std::endl;
        lcaFile.close();
        return EXIT_FAILURE;
      }
      lcas[std::make_tuple(tokens[0], tokens[1])] = common::Symbol(tokens[2]);
    }
  }
  lcaFile.close();

#ifndef NDEBUG
  std::cout << std::endl << "LCAs:  ";
  std::copy(lcas.begin(), lcas.end(), std::ostream_iterator<common::StrStrMap::value_type>(std::cout, "\t"));
  std::cout << std::endl;
#endif /* NDEBUG */

  std::ifstream set1File;
  set1File.open(args.set_1.c_str());
  if (!set1File.is_open()) {
    std::cerr << "Could not open file: " << args.set_1 << std::endl;
    return EXIT_FAILURE;
  }

#ifndef NDEBUG
  std::cout << "Reading the set1..." << std::endl;
#endif /* NDEBUG */

  common::Sequences seqs_1;
  while (!set1File.eof()) {
    std::getline(set1File, line);
#ifndef NDEBUG
    std::cout << "Read line: " << line << std::endl;
#endif /* NDEBUG */

    if (line != "") {
      common::StringVec cats;
      boost::split(cats, line, boost::is_any_of(","), boost::token_compress_on);
      seqs_1.push_back(cats);
    }
  }
  set1File.close();

#ifndef NDEBUG
  std::cout << std::endl << "1. Sequences:  ";
  std::copy(seqs_1.begin(), seqs_1.end(), std::ostream_iterator<common::StringVec>(std::cout, "\n"));
  std::cout << std::endl;
#endif /* NDEBUG */

  std::ifstream set2File;
  set2File.open(args.set_2.c_str());
  if (!set2File.is_open()) {
    std::cerr << "Could not open file: " << args.set_2 << std::endl;
    return EXIT_FAILURE;
  }

#ifndef NDEBUG
  std::cout << "Reading the set2..." << std::endl;
#endif /* NDEBUG */

  common::Sequences seqs_2;
  while (!set2File.eof()) {
    std::getline(set2File, line);
#ifndef NDEBUG
    std::cout << "Read line: " << line << std::endl;
#endif /* NDEBUG */

    if (line != "") {
      common::StringVec cats;
      boost::split(cats, line, boost::is_any_of(","), boost::token_compress_on);
      seqs_2.push_back(cats);
    }
  }
  set2File.close();

#ifndef NDEBUG
  std::cout << std::endl << "2. Sequences:  ";
  std::copy(seqs_2.begin(), seqs_2.end(), std::ostream_iterator<common::StringVec>(std::cout, "\n"));
  std::cout << std::endl;
#endif /* NDEBUG */

  alignment::AbstractDistanceMeasure *scoringScheme = new TreePathSimilarityMeasure(args.gap_penalty, euler_levels, euler_positions, lcas);
  alignment::SimilarityAlgorithm *similarity;

  if (args.alg == 1) {
    similarity = new alignment::SW(args.scores);
  } else {
    similarity = new alignment::NW(args.scores);
  }

  std::string outFile = args.results_dir + "/similarity-scores.dat";
  std::ofstream out(outFile.c_str(), std::ios::out);

  for (boost::uint32_t i = 0; i < seqs_1.size(); ++i) {
    #pragma omp parallel shared(std::cout, out, seqs_1, seqs_2, i, scoringScheme, similarity) default(none)
    {
      alignment::MemoryPool mem;
      std::vector<double> scores;

      #pragma omp for
      for (boost::uint32_t j = 0; j < seqs_2.size(); ++j) {
#ifndef NDEBUG
        std::cout << "i: ";
        std::copy(seqs_1[i].begin(), seqs_1[i].end(), std::ostream_iterator<std::string>(std::cout, " "));
        std::cout << std::endl;
        std::cout << "j: ";
        std::copy(seqs_2[j].begin(), seqs_2[j].end(), std::ostream_iterator<std::string>(std::cout, " "));
        std::cout << std::endl;
#endif /* NDEBUG */

        alignment::alignmentResult res = similarity->align(seqs_1[i], seqs_2[j], *scoringScheme, mem);
        // compute a normalised similarity measure
        double numerator = res.score * res.score;
        auto minSize = std::min(seqs_1[i].size(), seqs_2[j].size());
        double denominator = boost::lexical_cast<double>(minSize * minSize);
        double score = numerator/denominator;
        scores.push_back(score);

#ifndef NDEBUG
        std::cout << "Score: " << res.score << std::endl;
        std::cout << "Alignments: " << std::endl;
        std::copy(res.alignment[0].begin(), res.alignment[0].end(), std::ostream_iterator<std::string>(std::cout, " "));
        std::cout << std::endl;
#endif /* NDEBUG */
      }

      #pragma omp critical
      {
        for (std::vector<double>::iterator s_it = scores.begin(); s_it != scores.end(); ++s_it) {
          out << *s_it << std::endl;
        }
      }
    }
  }

  delete similarity;
  delete scoringScheme;

  return EXIT_SUCCESS;
}


src/main/CL.cc
// Copyright (C) 2015 Dominik Dahlem <Dominik.Dahlem@gmail.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

/** @file CL.cc
 * Implementation of the command-line parsing of the main routine
 *
 * @author Dominik Dahlem
 */
#if HAVE_CONFIG_H
# include <config.h>
#endif

#ifndef __STDC_CONSTANT_MACROS
# define __STDC_CONSTANT_MACROS
#endif /* __STDC_CONSTANT_MACROS */

#include <iostream>

#include <boost/cstdint.hpp>
#include <boost/filesystem.hpp>
namespace fs = boost::filesystem;

#include <boost/program_options/options_description.hpp>
#include <boost/program_options/parsers.hpp>
#include <boost/program_options/variables_map.hpp>
namespace po = boost::program_options;

#include "CL.hh"


CL::CL()
    : m_opt_desc(new po::options_description("Options"))
{
  // Declare the supported options.
  po::options_description opt_general("General Configuration");
  opt_general.add_options()
      (HELP.c_str(), "produce help message")
      (VERS.c_str(), "show the version")
      ;

  po::options_description opt_io("I/O Configuration");
  opt_io.add_options()
      (RESULTS_DIR.c_str(), po::value <std::string>()->default_value("./results"), "results directory.")
      (EULER_LEVELS.c_str(), po::value <std::string>()->default_value(""), "Filename of the vertex levels in the Euler Circuit.")
      (EULER_POSITIONS.c_str(), po::value <std::string>()->default_value(""), "Filename of the vertex positions in the Euler Circuit.")
      (LCA.c_str(), po::value <std::string>()->default_value(""), "Filename with the LCAs computed offline.")
      (SET_1.c_str(), po::value <std::string>()->default_value(""), "Filename of the source set.")
      (SET_2.c_str(), po::value <std::string>()->default_value(""), "Filename of the target set.")
      ;

  po::options_description opt_ha("Algorithm Configuration");
  opt_ha.add_options()
      (ALG.c_str(), po::value <boost::int32_t>()->default_value(1), "Algorithm: 1 - local alignment, 2 - global alignment.")
      (SCORES.c_str(), po::value <bool>()->default_value(0), "Compute just alignment scores, no backtracking.")
      (GAP_PENALTY.c_str(), po::value <double>()->default_value(1.33), "Gap penalty for the alignments.")
      ;

  m_opt_desc->add(opt_general);
  m_opt_desc->add(opt_io);
  m_opt_desc->add(opt_ha);
}


int CL::parse(int argc, char *argv[], args_t &p_args)
{
  po::variables_map vm;

  po::store(po::parse_command_line(argc, argv, (*m_opt_desc.get())), vm);
  po::notify(vm);

  if (vm.count(HELP)) {
    std::cout << (*m_opt_desc.get()) << std::endl;
    return EXIT_FAILURE;
  }

  if (vm.count(VERS)) {
    std::cout << PACKAGE_NAME << " " << PACKAGE_VERSION << std::endl;
    std::cout << argv[0] << std::endl;
    return EXIT_FAILURE;
  }

  if (vm.count(RESULTS_DIR.c_str())) {
    p_args.results_dir = vm[RESULTS_DIR.c_str()].as <std::string>();
  }

  if (vm.count(EULER_LEVELS.c_str())) {
    p_args.euler_levels = vm[EULER_LEVELS.c_str()].as <std::string>();
    if (!fs::exists(p_args.euler_levels)) {
      std::cerr << "The filename " << p_args.euler_levels << " with the Vertex levels in the Euler Circuit does not exist!" << std::endl;
      return EXIT_FAILURE;
    }
  }

  if (vm.count(EULER_POSITIONS.c_str())) {
    p_args.euler_positions = vm[EULER_POSITIONS.c_str()].as <std::string>();
    if (!fs::exists(p_args.euler_levels)) {
      std::cerr << "The filename " << p_args.euler_positions << " with the Vertex positions in the Euler Circuit does not exist!" << std::endl;
      return EXIT_FAILURE;
    }
  }

  if (vm.count(LCA.c_str())) {
    p_args.lca = vm[LCA.c_str()].as <std::string>();
    if (!fs::exists(p_args.lca)) {
      std::cerr << "The filename " << p_args.lca << " containing the LCAs does not exist!" << std::endl;
      return EXIT_FAILURE;
    }
  }

  if (vm.count(SET_1.c_str())) {
    p_args.set_1 = vm[SET_1.c_str()].as <std::string>();
    if (!fs::exists(p_args.set_1)) {
      std::cerr << "The filename " << p_args.set_1 << " containing the source set does not exist!" << std::endl;
      return EXIT_FAILURE;
    }
  }

  if (vm.count(SET_2.c_str())) {
    p_args.set_2 = vm[SET_2.c_str()].as <std::string>();
    if (!fs::exists(p_args.set_2)) {
      std::cerr << "The filename " << p_args.set_2 << " containing the target set does not exist!" << std::endl;
      return EXIT_FAILURE;
    }
  }

  if (vm.count(ALG.c_str())) {
    p_args.alg = vm[ALG.c_str()].as <boost::int32_t>();
  }

  if (vm.count(SCORES.c_str())) {
    p_args.scores = vm[SCORES.c_str()].as <bool>();
  }

  if (vm.count(GAP_PENALTY.c_str())) {
    p_args.gap_penalty = vm[GAP_PENALTY.c_str()].as <double>();
  }

  std::cout << argv[0] << " " << PACKAGE_VERSION << std::endl;
  std::cout << PACKAGE_NAME << std::endl;
  std::cout << p_args << std::endl;

  return EXIT_SUCCESS;
}
